

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>The import tool &mdash; Biobank 0.6.0_dev documentation</title>
    
    <link rel="stylesheet" href="../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.6.0_dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="Biobank 0.6.0_dev documentation" href="../index.html" />
    <link rel="up" title="Tools Usage" href="index.html" />
    <link rel="next" title="The query tool" href="query.html" />
    <link rel="prev" title="Tools Usage" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="query.html" title="The query tool"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Tools Usage"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Biobank 0.6.0_dev documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Tools Usage</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">The import tool</a><ul>
<li><a class="reference internal" href="#import-study">Import study</a></li>
<li><a class="reference internal" href="#import-individual">Import individual</a></li>
<li><a class="reference internal" href="#import-biosample">Import biosample</a></li>
<li><a class="reference internal" href="#import-biosample-container">Import biosample container</a></li>
<li><a class="reference internal" href="#import-device">Import device</a></li>
<li><a class="reference internal" href="#import-data-sample">Import data_sample</a></li>
<li><a class="reference internal" href="#import-data-object">Import data_object</a></li>
<li><a class="reference internal" href="#import-data-collection">Import data_collection</a></li>
<li><a class="reference internal" href="#import-diagnosis">Import diagnosis</a></li>
<li><a class="reference internal" href="#import-marker-alignment">Import marker_alignment</a></li>
<li><a class="reference internal" href="#import-markers-set">Import markers_set</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Tools Usage</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="query.html"
                        title="next chapter">The query tool</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/usage/import.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-bl.vl.app.importer.main">
<span id="the-import-tool"></span><span id="import-tool"></span><h1>The import tool<a class="headerlink" href="#module-bl.vl.app.importer.main" title="Permalink to this headline">¶</a></h1>
<p>The purpose of this tool is to import externally generated data into
the KB. Import operations can almost always be described as:</p>
<blockquote>
<div><ul class="simple">
<li>reading a tsv input file with one column named <strong>label</strong>, a unique
id for the specific object defined and other fields. When the
specific object is linked to another, e.g., a blood sample to an
individual, there will be one column named <strong>source</strong> with the VID
(the unique id assigned by the KB to the object) of an object to
which this object should be linked to;</li>
<li>saving data read from the tsv file to the KB, plus other data
provided as parameters. For each input record, the program
generates the corresponding object as well as any relevant action
that links it to its source;</li>
<li>outputting an object mapping tsv file with four columns: study,
label, object_type and VID, where object_type is the KB object type.</li>
</ul>
</div></blockquote>
<p>The user is responsible for mapping labels to VIDs: in most cases you
should be able to use <tt class="docutils literal"><span class="pre">kb_query</span> <span class="pre">map_vid</span></tt> for this.</p>
<span class="target" id="module-bl.vl.app.importer.study"></span><div class="section" id="import-study">
<h2>Import study<a class="headerlink" href="#import-study" title="Permalink to this headline">¶</a></h2>
<p>A study represents a general context. It is characterized by the
following fields:</p>
<div class="highlight-python"><pre>label  description
ASTUDY A textual description of ASTUDY, no tabs please.</pre>
</div>
<p>The description column is optional. The study sub-operation will read
in a tsv files with the above information and output the VIDs of the
newly created study objects.</p>
</div>
<span class="target" id="module-bl.vl.app.importer.individual"></span><div class="section" id="import-individual">
<h2>Import individual<a class="headerlink" href="#import-individual" title="Permalink to this headline">¶</a></h2>
<p>An individual is characterized by the following fields:</p>
<div class="highlight-python"><pre>study label gender   father mother
xxx   id2   male     None   None
xxx   id2   male     id4    id5</pre>
</div>
<p>where gender can be either male or female; father and mother can be
either the string &#8216;None&#8217; or the label of an individual in the same
study.</p>
<p>Individuals are the only &#8216;bio&#8217; objects that can be loaded
independently from the KB&#8217;s current content (if their parents are
&#8216;None&#8217;). The study label should, however, correspond to a previously
loaded study.</p>
<p><strong>NOTE:</strong> The current implementation does not support cross-study
kinship.</p>
</div>
<span class="target" id="module-bl.vl.app.importer.biosample"></span><div class="section" id="import-biosample">
<h2>Import biosample<a class="headerlink" href="#import-biosample" title="Permalink to this headline">¶</a></h2>
<p>A biosample record will have, at least, the following fields:</p>
<div class="highlight-python"><pre>label     source
I001-bs-2 V932814892
I002-bs-2 V932814892
I003-bs-2 None</pre>
</div>
<p>Where label is the label of the biosample container. If a &#8216;None&#8217; value
has been passed in the source column, the biosample will be imported
as a new unlinked object into the biobanks. Another example, this time
involving DNA samples:</p>
<div class="highlight-python"><pre>label    source     used_volume current_volume activation_date
I001-dna V932814899 0.3         0.2            17/03/2007
I002-dna V932814900 0.22        0.2            21/01/2004</pre>
</div>
<p>A special case is when records refer to biosamples contained in plate
wells. In this case, an additional column must be present with the VID
of the corresponding TiterPlate object. For instance:</p>
<div class="highlight-python"><pre>plate  label source
V39030 A01   V932814892
V39031 A02   V932814893
V39032 A03   V932814894</pre>
</div>
<p>where the label column is now the label of the well position.</p>
<p>If row and column (optional) are provided, the program will use them;
if they are not provided, it will infer them from label (e.g., J01 -&gt;
row=10, column=1). Missing labels will be generated as:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="s">&#39;</span><span class="si">%s%03d</span><span class="s">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">chr</span><span class="p">(</span><span class="n">row</span><span class="o">+</span><span class="nb">ord</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="n">column</span><span class="p">)</span>
</pre></div>
</div>
<p>A badly formed label will result in the rejection of the record; the
same will happen if label, row and column are inconsistent. The well
will be filled by current_volume material produced by removing
used_volume material taken from the bio material contained in the
vessel identified by source. row and column are base 1.</p>
<p>If the sample is a IlluminaBeadChipArray the plate column used in the
PlateWell case will become a illumina_array column and a new column, named
bead_chip_assay_type, is required.</p>
<blockquote>
<div>illumina_array  label   source   bead_chip_assay_type
V1351235        R01C01  V412441  HUMANEXOME_12V1_B
V1351235        R01C02  V351151  HUMANEXOME_12V1_B
V1351235        R02C01  V345115  HUMANEXOME_12V1_B</div></blockquote>
</div>
<span class="target" id="module-bl.vl.app.importer.samples_container"></span><div class="section" id="import-biosample-container">
<h2>Import biosample container<a class="headerlink" href="#import-biosample-container" title="Permalink to this headline">¶</a></h2>
<p>A container record will have the following fields:</p>
<div class="highlight-python"><pre>label         container_status  creation_date
A_CONTAINER   USABLE            13/02/2012
B_CONTAINER   INSTOCK           12/01/2001
C_CONTAINER   USABLE            25/04/2012
....</pre>
</div>
<p>the creation_date column is optional, if not specified current date
will be set as the object&#8217;s creation date, also the container_status
column can be optional if this values is passed as input parameter.</p>
<p>When importing new containers, special fields can be included in the
CSV file depending on the type of the objects that you want to
import.</p>
<p>For TITER PLATES objects the syntax can be the following:</p>
<div class="highlight-python"><pre>label         barcode    container_status  rows  columns
A_TITERPLATE  XXYYZZ111  INSTOCK           8     12
B_TITERPLATE  XXYYZZ112  INSTOCK           8     12
C_TITERPLATE  XXYYZZ113  READY             8     12
....</pre>
</div>
<p>rows and columns values can be optional if these values are passed as
input parameters, barcode column is optional.</p>
<p>For ILLUMINA ARRAY OF ARRAYS objects the syntax can be the following</p>
<blockquote>
<div>label        barcode    container_status   rows   columns   illumina_array_type   illumina_array_class   illumina_assay_type
A_ILLARRAY   XXYYZZ111  INSTOCK            4      2         BeadChip_12x1Q        Slide                  Infinium_HD
B_ILLARRAY   XXYYZZ112  INSTOCK            4      2         BeadChip_12x1Q        Slide                  Infinium_HD
C_ILLARRAY   XXYYZZ113  INSTOCK            4      2         BeadChip_12x1Q        Slide                  Infinium_HD</div></blockquote>
<p>rows, columns, illumina_array_type, illumina_array_class and illumina_assay_type
can be optional if these values are passed as input parameters, barcode column
is optional.</p>
<p>For FLOW CELL objects the syntax can be the following:</p>
<div class="highlight-python"><pre>label       barcode     container_status  number_of_slots
A_FLOWCELL  XXYYZZ221   INSTOCK           8
B_FLOWCELL  XXYYZZ222   INSTOCK           8
C_FLOWCELL  XXYYZZ223   INSTOCK           8
....</pre>
</div>
<p>number_of_slots column can be optional if this value is passed as
input paramter, barcode column is optional.</p>
<p>For LANE objects the syntax can be the following:</p>
<div class="highlight-python"><pre>flow_cell   slot  container_status
V112441441  1     INSTOCK
V112441441  2     INSTOCK
V112441441  3     INSTOCK
V351145519  1     INSTOCK
V351145519  2     INSTOCK
....</pre>
</div>
<p>for Lane objects, no label column has to be provided, the importer
will automatically calculate the labels for each imported object.</p>
</div>
<span class="target" id="module-bl.vl.app.importer.device"></span><div class="section" id="import-device">
<h2>Import device<a class="headerlink" href="#import-device" title="Permalink to this headline">¶</a></h2>
<p>Will read in a tsv file with the following columns (multiple spaces
represent tabs):</p>
<div class="highlight-python"><pre>device_type  label   maker       model                        release
Scanner      pula01  Affymetrix  GeneChip Scanner 3000        7G
Chip         chip01  Affymetrix  Genome-Wide Human SNP Array  6.0</pre>
</div>
<p>All devices have a type, a label, a maker, a model and a release;
optional fields are &#8216;barcode&#8217; and &#8216;location&#8217;.</p>
</div>
<span class="target" id="module-bl.vl.app.importer.data_sample"></span><div class="section" id="import-data-sample">
<h2>Import data_sample<a class="headerlink" href="#import-data-sample" title="Permalink to this headline">¶</a></h2>
<p>Will read in a tsv file with the following columns:</p>
<div class="highlight-python"><pre>study  label source  device  device_type scanner options             status
ASTUDY foo01 V039090 V093090 Chip        v099020 celID=0009099090    USABLE
ASTUDY foo02 V039090 V099022 Scanner     v099022 conf1=...,conf2=... UNKNOWN
...</pre>
</div>
<p>and instantiate a specialized DataSample-derived class for each line.</p>
<p>In the above example, the first data sample has been obtained by using
chip V093090 on scanner V099020, while the second one has been
obtained using a direct scanning technology, e.g., an Illumina HiSeq
2000. The optional scanner column, the vid of the scanner device, is
used in cases, such as Affymetrix genotyping, where it is relevant.</p>
<p>The general strategy is to decide what DataSample subclasses should be
instantiated by retrieving the Device and look at its maker, model,
release attributes. The optional data_sample_type column overrides
all automatic decisions.</p>
<p>It is also possible to import DataSample(s) that are the results of
processing other DataSample(s). Here is an example:</p>
<div class="highlight-python"><pre>study  label source  device  device_type     options             status
ASTUDY foo01 V039090 V099021 SoftwareProgram conf1=...,conf2=... USABLE
ASTUDY foo02 V039090 V099021 SoftwareProgram conf1=...,conf2=... USABLE
...</pre>
</div>
<p>A special case is the GenotypeDataSample, where it is mandatory to
assign a SNPMarkerSet by listing its VID in the markers_set column. As
an example:</p>
<div class="highlight-python"><pre>study  label source device   device_type     data_sample_type   markers_set status
ASTUDY foo01 V039090 V099021 SoftwareProgram GenotypeDataSample V020202      USABLE
ASTUDY foo02 V039090 V099021 SoftwareProgram GenotypeDataSample V020202      USABLE
...</pre>
</div>
</div>
<span class="target" id="module-bl.vl.app.importer.data_object"></span><div class="section" id="import-data-object">
<h2>Import data_object<a class="headerlink" href="#import-data-object" title="Permalink to this headline">¶</a></h2>
<p>Will read in a tsv file with at least the following columns:</p>
<div class="highlight-python"><pre>path data_sample size sha1
/share/fs/v039303.cel V02902 39090 SHA1SHA1
...</pre>
</div>
<p>and will instantiate, for each row, a DataObject that represents a
physical instance of a given DataSample.</p>
<p>Records that point to an unknown data sample vid will be noisily
ignored. The same will happen to records that have the same path of a
previously seen data object.</p>
</div>
<span class="target" id="module-bl.vl.app.importer.data_collection"></span><div class="section" id="import-data-collection">
<h2>Import data_collection<a class="headerlink" href="#import-data-collection" title="Permalink to this headline">¶</a></h2>
<p>Will read in a tsv file with the following columns:</p>
<div class="highlight-python"><pre>study    label data_sample
BSTUDY   dc-01 V0390290
BSTUDY   dc-01 V0390291
BSTUDY   dc-02 V0390292
BSTUDY   dc-02 V390293
...</pre>
</div>
<p>This will create new DataCollection(s), whose label is defined by the
label column, and link to it, using DataCollectionItem objects,
the DataSample object identified by data_sample (a VID).</p>
<p>Records that point to an unknown DataSample will abort the data
collection loading. Previously seen collections will be noisily
ignored. It is not legal to use the importer to add items to a
previously known collection.</p>
</div>
<span class="target" id="module-bl.vl.app.importer.diagnosis"></span><div class="section" id="import-diagnosis">
<h2>Import diagnosis<a class="headerlink" href="#import-diagnosis" title="Permalink to this headline">¶</a></h2>
<p>Will read in a tsv file with the following columns:</p>
<div class="highlight-python"><pre>study  individual timestamp      diagnosis
ASTUDY V899       1310057541608  icd10-cm:E10
ASTUDY V899       1310057541608  icd10-cm:G35
ASTYDY V1806      1310057541608  exclusion-problem_diagnosis
...</pre>
</div>
<p>where the diagnosis column contains <a class="reference external" href="http://www.openehr.org">openEHR</a> diagnosis codes.</p>
</div>
<span class="target" id="module-bl.vl.app.importer.marker_alignment"></span><div class="section" id="import-marker-alignment">
<h2>Import marker_alignment<a class="headerlink" href="#import-marker-alignment" title="Permalink to this headline">¶</a></h2>
<p>Adds marker alignment info to a marker set.  The input tsv file is
structured as follows:</p>
<div class="highlight-python"><pre>marker_vid chromosome pos      strand allele copies
V0909090   10         82938938 True   A      1
V0909091   1          82938999 True   A      2
V0909092   1          82938938 True   B      2
...</pre>
</div>
<p>An alignment record is required for each marker in the set.  Missing
alignment info must be filled by dummy records:</p>
<div class="highlight-python"><pre>V0909093   None       None     None   None   0</pre>
</div>
<p>The pos fields is relative to 5&#8217;: if the marker has been aligned on
the other strand, it is the responsibility of the aligner app to
report the actual distance from 5&#8217;, while, at the same time,
registering that the SNP has actually been aligned on the other
strand. The chromosome field is an integer field with values in the
<tt class="docutils literal"><span class="pre">[1,</span> <span class="pre">26]</span></tt> range, with 23-26 representing, respectively, the X
chromosome, the Y chromosome, the pseudoautosomal regions (XY) and the
mitochondrial DNA (MT).</p>
</div>
<span class="target" id="module-bl.vl.app.importer.markers_set"></span><div class="section" id="import-markers-set">
<h2>Import markers_set<a class="headerlink" href="#import-markers-set" title="Permalink to this headline">¶</a></h2>
<p>FIXME &#8212; this is out phase with the current classes</p>
<p>A SNPMarkersSet represents an ordered list of markers where the order
usually comes from aligning the SNP markers against a reference
genome. Within Biobank, different genotyping technologies are mapped
to different SNPMarkersSet instances.</p>
<p>In more detail, a marker set is defined by:</p>
<blockquote>
<div><ul class="simple">
<li>for the whole set:<ul>
<li>maker: the name of the organization that has defined the
SNPMarkersSet, e.g., &#8216;CRS4&#8217;</li>
<li>model: the specific &#8216;model&#8217; of the SNPMarkersSet, e.g.,
&#8216;AffymetrixGenome6.0&#8217;</li>
<li>release: a string that identifies this specific instance, e.g.,
&#8216;aligned_on_human_g1k_v37&#8217;</li>
</ul>
</li>
<li>for each marker in the set:<ul>
<li>label: a descriptive string, usually assigned by the manufacturer
and unique within the specific genotyping platform</li>
<li>mask: a string that describes the DNA structure of the marker, in
the &lt;FLANK&gt;[A/B]&lt;FLANK&gt; format, e.g., ACGTCCAC[A/G]ACTAGCTA.  All
input masks are expected to be in the TOP Illumina convention, if
the Illumina strand detection algorithm yields a result (see
<a class="reference internal" href="../dev_manual/api_ref.html#bl.vl.utils.snp.convert_to_top" title="bl.vl.utils.snp.convert_to_top"><tt class="xref py py-func docutils literal"><span class="pre">convert_to_top()</span></tt></a>).</li>
<li>index: the position of the marker within the marker list</li>
<li>allele_flip: False if the alleles are in the same order as
recorded in the manufacturer&#8217;s data sheet</li>
</ul>
</li>
</ul>
</div></blockquote>
<p>Set info is provided through command line parameters, while per-marker
info must be listed in the tab-separated input file.  For instance:</p>
<div class="highlight-python"><pre>label  mask               index  allele_flip
SNP-1  ACGTCC[A/G]ACTAGC  0      False
SNP-2  CGATCG[T/C]ACACTG  1      False
SNP-3  TGACTA[T/G]TAGCGA  2      True
...</pre>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="query.html" title="The query tool"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Tools Usage"
             >previous</a> |</li>
        <li><a href="../index.html">Biobank 0.6.0_dev documentation</a> &raquo;</li>
          <li><a href="index.html" >Tools Usage</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, CRS4.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>